tools/codegen/model_generator/parser.go

```diff
@@ -0,0 +1,564 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Code in this class is mostly a copy from https://github.com/golang/protobuf/blob/master/protoc-gen-go/generator/generator.go
+TODO:
+-  Handle name collision. When multiple files in the FileDescriptorSet have the same name, the protoc-gen-go do name
+   substitution before printing the package name. Our generated code that might reference those packages should also
+   use the substituted name that protoc-gen-go would use to name that package.
+-  Handle go_package option. Currently the name of the package is computed just based on the file path.
+-  Handle Oneofs. The protoc-gen-go have special logic (removed from this file, but present in the protoc-gen-go).
+   Our generated interface code might also need that special logic if the Instance object wants to reference a
+   oneof from the generated mytemplate.pb.go
+-  Handle Extensions. Should we allow Constructor message to have extension fields. Extensions have are a complete
+   different handling in protoc-gen-go. So far I am what it means for our generated code.
+-  Accept all the other parameters that protoc-gen-go takes in our generator. Our codegen should use those
+   paramters for it's codegen and pass to the protoc-gen-go plugin (invoking another protoc process).
+*/
+
+package model_generator
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+	"path"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
```

Comment:
i think we probably want to run bin/fmt.sh on these files. should fix up the import blocks.